File: service/lib/agama/registration.rb
# frozen_string_literal: true

# Copyright (c) [2023-2025] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "fileutils"
require "yast"
require "ostruct"
require "suse/connect"
require "y2packager/new_repository_setup"
require "y2packager/resolvable"
require "yast2/execute"

require "agama/cmdline_args"
require "agama/errors"
require "agama/registered_addon"
require "agama/ssl/certificate"
require "agama/ssl/errors"
require "agama/ssl/fingerprint"
require "agama/ssl/storage"

Yast.import "Arch"
Yast.import "Pkg"

module Agama
  # Handles everything related to registration of system to SCC, RMT or similar.
  class Registration
    include Yast::I18n

    # NOTE: identical and keep in sync with Software::Manager::TARGET_DIR
    TARGET_DIR = "/run/agama/zypp"
    private_constant :TARGET_DIR

    # FIXME: it should use TARGET_DIR instead of "/", but connect failed to read it even
    # if fs_root passed as client params. Check with SCC guys why.
    GLOBAL_CREDENTIALS_PATH = File.join("/",
      SUSE::Connect::YaST::GLOBAL_CREDENTIALS_FILE)
    private_constant :GLOBAL_CREDENTIALS_PATH

    DEFAULT_REGISTRATION_URL = "https://scc.suse.com"
    private_constant :DEFAULT_REGISTRATION_URL

    # Code used for registering the product.
    #
    # @return [boolean] true if base product is already registered
    attr_reader :registered

    # Code used for registering the product.
    #
    # @return [String, nil] nil if the product is not registered yet.
    attr_reader :reg_code

    # Email used for registering the product.
    #
    # @return [String, nil]
    attr_reader :email

    # List of already registered addons
    #
    # @return [Array<RegisteredAddon>]
    attr_reader :registered_addons

    # Overwritten default registration URL
    #
    # @return [String, nil]
    attr_accessor :registration_url

    # @param software_manager [Agama::Software::Manager]
    # @param logger [Logger]
    def initialize(software_manager, logger)
      @software = software_manager
      @logger = logger
      @services = []
      @credentials_files = []
      @registered_addons = []
      @registration_url = registration_url_from_cmdline
      @registered = false
    end

    # Registers the selected product.
    #
    # @raise [
    #   SocketError|Timeout::Error|SUSE::Connect::ApiError|
    #   SUSE::Connect::MissingSccCredentialsFile|SUSE::Connect::MissingSccCredentialsFile|
    #   OpenSSL::SSL::SSLError|JSON::ParserError|Agama::Software::AddServiceError
    # ]
    #
    # @param code [String] Registration code.
    # @param email [String] Email for registering the product.
    def register(code, email: "")
      return if product.nil? || registered

      catch_registration_errors do
        reg_params = connect_params(token: code, email: email)

        # hide the registration code in the logs
        log_params = reg_params.dup
        log_params[:token] = "<REG_CODE>" if log_params[:token] != ""
        @logger.info "Registering system #{target_distro}: #{log_params.inspect}"

        @login, @password = SUSE::Connect::YaST.announce_system(reg_params, target_distro)
        # write the global credentials
        # TODO: check if we can do it in memory for libzypp
        SUSE::Connect::YaST.create_credentials_file(@login, @password, GLOBAL_CREDENTIALS_PATH)

        activate_params = {}
        @logger.info "Activating product #{base_target_product.inspect}"
        service = SUSE::Connect::YaST.activate_product(base_target_product, activate_params, email)
        process_service(service)
        @registered = true
      end
    ensure
      @reg_code = code
      @email = email
      run_on_change_callbacks
    end

    def register_addon(name, version, code)
      catch_registration_errors do
        register_version = if version.empty?
          # version is not specified, find it automatically
          find_addon_version(name)
        else
          # use the explicitly required version
          version
        end

        if @registered_addons.any? { |a| a.name == name && a.version == register_version }
          @logger.info "Addon #{name}-#{register_version} already registered, skipping registration"
          return
        end

        @logger.info "Registering addon #{name}-#{register_version}"
        # do not log the code, but at least log if it is empty
        @logger.info "Using empty registration code" if code.empty?

        target_product = OpenStruct.new(
          arch:       Yast::Arch.rpm_arch,
          identifier: name,
          version:    register_version
        )
        activate_params = { token: code }
        service = SUSE::Connect::YaST.activate_product(target_product, activate_params, @email)
        process_service(service)

        @registered_addons << RegisteredAddon.new(name, register_version, !version.empty?, code)
        # select the products to install
        @software.addon_products(find_addon_products)

        run_on_change_callbacks
      end
    end

    # Deregisters the selected product.
    #
    # It uses the registration code and email passed to {#register}.
    #
    # @raise [
    #   SocketError|Timeout::Error|SUSE::Connect::ApiError|
    #   SUSE::Connect::MissingSccCredentialsFile|SUSE::Connect::MissingSccCredentialsFile|
    #   OpenSSL::SSL::SSLError|JSON::ParserError
    # ]
    def deregister
      return unless registered

      @services.each do |service|
        Y2Packager::NewRepositorySetup.instance.services.delete(service.name)
        @software.remove_service(service)
      end

      # reset
      @software.addon_products([])
      @services = []
      @available_addons = nil

      reg_params = connect_params(token: reg_code, email: email)
      SUSE::Connect::YaST.deactivate_system(reg_params)
      FileUtils.rm(GLOBAL_CREDENTIALS_PATH) # connect does not remove it itself
      @credentials_files.each do |credentials_file|
        FileUtils.rm(File.join(TARGET_DIR, credentials_path(credentials_file)))
      end
      @credentials_files = []

      @registered = false
      @reg_code = nil
      @email = nil
      @registered_addons = []

      run_on_change_callbacks
    end

    # Copies configuration and credentials files to the target system.
    #
    # The configuration file is copied only if a registration URL was given.
    def finish
      return unless registered

      files = [[
        GLOBAL_CREDENTIALS_PATH, File.join(Yast::Installation.destdir, GLOBAL_CREDENTIALS_PATH)
      ]]
      @credentials_files.each do |credentials_file|
        files << [
          File.join(TARGET_DIR, credentials_path(credentials_file)),
          File.join(Yast::Installation.destdir, credentials_path(credentials_file))
        ]
      end

      if registration_url
        SUSE::Connect::YaST.write_config("url" => registration_url)
        files << [
          SUSE::Connect::Config::DEFAULT_CONFIG_FILE,
          File.join(Yast::Installation.destdir, SUSE::Connect::Config::DEFAULT_CONFIG_FILE)
        ]
      end

      files.each do |src_dest|
        FileUtils.cp(*src_dest)
      end

      copy_certificates_to_target
    end

    # Get the available addons for the specified base product.
    #
    # @note The result is bound to the registration code used for the base product, the result
    # might be different for different codes. E.g. the Alpha/Beta extensions might or might not
    # be included in the list.
    def available_addons
      return @available_addons if @available_addons

      @available_addons = SUSE::Connect::YaST.show_product(base_target_product,
        connect_params).extensions
      @logger.info "Available addons: #{available_addons.inspect}"
      @available_addons
    end

    # Callbacks to be called when registration changes (e.g., a different product is selected).
    def on_change(&block)
      @on_change_callbacks ||= []
      @on_change_callbacks << block
    end

  private

    # @return [Agama::Software::Manager]
    attr_reader :software

    # Currently selected product.
    #
    # @return [Agama::Software::Product, nil]
    def product
      software.product
    end

    def copy_certificates_to_target
      cert_file = SSL::Certificate.default_certificate_path
      return unless File.exist?(cert_file) # no certificate imported?

      # copy the imported certificate
      @logger.info "Copying SSL certificate (#{cert_file}) to the target system..."
      cert_target_file = File.join("/mnt",
        SUSE::Connect::YaST::SERVER_CERT_FILE)
      ::FileUtils.mkdir_p(File.dirname(cert_target_file))
      ::FileUtils.cp(cert_file, cert_target_file)

      # update the certificate links
      cmd = SUSE::Connect::YaST::UPDATE_CERTIFICATES
      @logger.info "Updating certificate links (#{cmd})..."
      # beware that Yast::Execute.on_target! does not work here due
      # to not changed SCR root when registration finish is executed.
      # So do chroot explicitelly.
      Yast::Execute.locally!(cmd, chroot: "/mnt")
    end

    # Product name expected by SCC.
    #
    # @return [String] E.g., "ALP-Dolomite-1-x86_64".
    def target_distro
      v = product.version.to_s.split(".").first || "1"
      "#{product.id}-#{v}-#{Yast::Arch.rpm_arch}"
    end

    def run_on_change_callbacks
      @on_change_callbacks&.map(&:call)
    end

    # taken from https://github.com/yast/yast-registration/blob/master/src/lib/registration/url_helpers.rb#L109
    def credentials_from_url(url)
      parsed_url = URI(url)
      params = URI.decode_www_form(parsed_url.query).to_h

      params["credentials"]
    rescue StandardError
      # if something goes wrong try to continue like if there is no credentials param
      nil
    end

    def credentials_path(file)
      File.join(SUSE::Connect::YaST::DEFAULT_CREDENTIALS_DIR, file)
    end

    # Returns the arguments to connect to the registration server
    #
    # @param params [Hash] additional parameters (e.g., email and token)
    # @return [Hash]
    def connect_params(params = {})
      default_params = {}
      default_params[:language] = http_language if http_language
      default_params[:url] = registration_url || DEFAULT_REGISTRATION_URL
      default_params[:verify_callback] = verify_callback
      default_params.merge(params)
    end

    def http_language
      lang = Yast::WFM.GetLanguage
      return nil if ["POSIX", "C"].include?(lang)

      # remove the encoding suffix (e.g. ".UTF-8")
      lang = lang.sub(/\..*$/, "")

      # replace Linux locale separator "_" by the HTTP separator "-", downcase the country name
      # see https://www.rfc-editor.org/rfc/rfc9110.html#name-accept-language
      lang.tr!("_", "-")
      lang.downcase!

      lang
    end

    # returns SSL verify callback
    def verify_callback
      lambda do |verify_ok, context|

        # we cannot raise an exception with details here (all exceptions in
        # verify_callback are caught and ignored), we need to store the error
        # details in a global instance
        store_ssl_error(context) unless verify_ok

        verify_ok
      rescue StandardError => e
        @logger.error "Exception in SSL verify callback: #{e.class}: #{e.message} : #{e.backtrace}"
        # the exception will be ignored, but reraise anyway...
        raise e

      end
    end

    def store_ssl_error(context)
      @logger.error "SSL verification failed: #{context.error}: #{context.error_string}"
      SSL::Errors.instance.ssl_error_code = context.error
      SSL::Errors.instance.ssl_error_msg = context.error_string
      SSL::Errors.instance.ssl_failed_cert =
        context.current_cert ? SSL::Certificate.load(context.current_cert) : nil
    end

    def catch_registration_errors(&block)
      # import the SSL certificate just once to avoid an infinite loop
      certificate_imported = false
      begin
        # reset the previous SSL errors
        Agama::SSL::Errors.instance.reset

        block.call

        true
      rescue OpenSSL::SSL::SSLError => e
        @logger.error "OpenSSL error: #{e}"
        should_retry = handle_ssl_error(e, certificate_imported)
        puts "handle ssl error #{should_retry}"
        if should_retry
          certificate_imported = true
          SSL::Errors.instance.ssl_failed_cert.import
          retry
        end
        raise e
      end
    end

    # @return [Boolean]
    def handle_ssl_error(_error, certificate_imported)
      return false if certificate_imported

      cert = SSL::Errors.instance.ssl_failed_cert
      return false unless cert

      # Import certificate if it matches predefined fingerprint.
      return true if SSL::Storage.instance.fingerprints.any? { |f| cert.match_fingerprint?(f) }

      error_code = SSL::Errors.instance.ssl_error_code
      return false unless SSL::ErrorCodes::IMPORT_ERROR_CODES.include?(error_code)

      question = certificate_question(cert)
      questions_client = Agama::DBus::Clients::Questions.new(logger: @logger)
      questions_client.ask(question) { |c| c.answer == :trust }
    end

    # @param certificate [Agama::SSL::Certificate]
    # @return [Agama::Question]
    def certificate_question(certificate)
      question_data = {
        "url"                => registration_url || DEFAULT_REGISTRATION_URL,
        "issuer_name"        => certificate.issuer_name,
        "issue_date"         => certificate.issued_on,
        "expiration_date"    => certificate.expires_on,
        "sha1_fingerprint"   => certificate.fingerprint(SSL::Fingerprint::SHA1).value,
        "sha256_fingerprint" => certificate.fingerprint(SSL::Fingerprint::SHA256).value
      }.filter { |_, v| !v.nil? }

      question_text = _(
        "Trying to import a self signed certificate. Do you want to trust it and register the " \
        "product?"
      )

      Agama::Question.new(
        qclass:         "registration.certificate",
        text:           question_text,
        options:        [:trust, :reject],
        default_option: :reject,
        data:           question_data
      )
    end

    # Returns the URL of the registration server
    #
    # At this point, it just checks the kernel's command-line.
    #
    # @return [String, nil]
    def registration_url_from_cmdline
      cmdline_args = CmdlineArgs.read
      cmdline_args.data["register_url"]
    end

    # process a newly added service, create the credentials file and add the service to libzypp
    def process_service(service)
      @services << service
      credentials_file = credentials_from_url(service.url)
      if credentials_file
        @credentials_files << credentials_file
        # addons use the same SCC credentials as the base product
        SUSE::Connect::YaST.create_credentials_file(@login, @password,
          File.join(TARGET_DIR, credentials_path(credentials_file)))
      end
      Y2Packager::NewRepositorySetup.instance.add_service(service.name)
      @software.add_service(service)
    end

    # Find all addon products
    #
    # @return [Array<String>] names of the products
    def find_addon_products
      # find all repositories for registered addons (their services)
      addon_repos = @services.reduce([]) do |acc, service|
        # skip the first service, it belongs to the base product
        next acc if service == @services.first

        acc.concat(service_repos(service))
      end

      # find all products from those repositories
      products = Y2Packager::Resolvable.find(kind: :product)
      products.select! do |product|
        addon_repos.any? { |addon_repo| product.source == addon_repo["SrcId"] }
      end

      products.map!(&:name)
      @logger.info "Addon products to install: #{products}"

      products
    end

    # Find all repositories belonging to a service.
    #
    # @param product_service [OpenStruct] repository service from suseconnect
    #
    # @return [Array<Hash>] repository data as returned by the Pkg.SourceGeneralData
    #   call, additionally with the "SrcId" key
    def service_repos(product_service)
      @logger.info "product_service: #{product_service.inspect}"
      repo_data = Yast::Pkg.SourceGetCurrent(false).map { |repo| repository_data(repo) }

      service_name = product_service.name
      # select only repositories belonging to the product services
      repos = repo_data.select { |repo| service_name == repo["service"] }
      @logger.info "Service #{service_name.inspect} repositories: #{repos}"

      repos
    end

    # Get repository data
    # @param [Fixnum] repo repository ID
    # @return [Hash] repository properties, including the repository ID ("SrcId" key)
    def repository_data(repo)
      data = Yast::Pkg.SourceGeneralData(repo)
      data["SrcId"] = repo
      data
    end

    # Find the version for the specified addon, if none if multiple addons with the same name
    # are found an exception is thrown.
    #
    # @return [String] the addon version, e.g. "16.0"
    def find_addon_version(name)
      raise Errors::Registration::ExtensionNotFound, name unless available_addons

      requested_addons = available_addons.select { |a| a.identifier == name }
      case requested_addons.size
      when 0
        raise Errors::Registration::ExtensionNotFound, name
      when 1
        requested_addons.first.version
      else
        raise Errors::Registration::MultipleExtensionsFound.new(name,
          requested_addons.map(&:version))
      end
    end

    # Construct the base product data for sending to the server
    def base_target_product
      OpenStruct.new(
        arch:       Yast::Arch.rpm_arch,
        identifier: product.id,
        version:    product.version || "1.0"
      )
    end
  end
end

File: rust/agama-software/src/model/registration.rs
// Copyright (c) [2025] SUSE LLC
//
// All Rights Reserved.
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 2 of the License, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
// more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, contact SUSE LLC.
//
// To contact SUSE LLC about this file by physical or electronic mail, you may
// find current contact information at www.suse.com.

//! This module implements support for registering a system.
//!
//! It interacts with SUSEConnect-ng (using the [suseconnect_agama] crate) to register
//! the system and its add-ons and with libzypp (through [zypp_agama]) to add the
//! corresponding services to `libzypp`.

use agama_security as security;
use agama_utils::{
    actor::Handler,
    api::software::{AddonInfo, AddonRegistration, RegistrationInfo},
    arch::Arch,
    helpers::copy_dir_all,
};
use camino::Utf8PathBuf;
use openssl::x509::X509;
use suseconnect_agama::{self, ConnectParams, Credentials};
use url::Url;

use crate::state::Addon;

#[derive(thiserror::Error, Debug)]
pub enum RegistrationError {
    #[error(transparent)]
    Registration(#[from] suseconnect_agama::Error),
    #[error("Failed to add the service {0}: {1}")]
    AddService(String, #[source] zypp_agama::ZyppError),
    #[error("Failed to refresh the service {0}: {1}")]
    RefreshService(String, #[source] zypp_agama::ZyppError),
    #[error("Failed to copy file {0}: {1}")]
    IO(String, #[source] std::io::Error),
    #[error(transparent)]
    Security(#[from] security::service::Error),
}

type RegistrationResult<T> = Result<T, RegistrationError>;

/// Represents a registered system.
///
/// It is used to activate products and add the corresponding services.
/// It is created from a [RegistrationBuilder].
#[derive(Debug)]
pub struct Registration {
    root_dir: Utf8PathBuf,
    product: String,
    version: String,
    arch: Arch,
    // The connection parameters are kept because they are needed by the
    // `to_registration_info` function.
    connect_params: ConnectParams,
    creds: Credentials,
    services: Vec<suseconnect_agama::Service>,
    // Holds the addons information because the status cannot be obtained from SCC yet
    // (e.g., whether and add-on is register or its registration code).
    addons: Vec<Addon>,
    // Holds all config files it created, later it will be copied to target system
    config_files: Vec<Utf8PathBuf>,
}

impl Registration {
    pub fn builder(root_dir: Utf8PathBuf, product: &str, version: &str) -> RegistrationBuilder {
        RegistrationBuilder::new(root_dir, product, version)
    }

    /// Registers the given add-on.
    ///
    /// * `zypp`: zypp instance.
    /// * `addon`: add-on to register.
    pub fn register_addon(
        &mut self,
        zypp: &zypp_agama::Zypp,
        addon: &Addon,
    ) -> RegistrationResult<()> {
        // Use the product's version as default.
        let version = addon.version.clone().unwrap_or(self.version.clone());
        let code = addon.code.as_ref().map(|c| c.as_str());
        self.activate_product(&zypp, &addon.id, &version, code)?;
        self.addons.push(addon.clone());
        Ok(())
    }

    /// Determines whether an add-on is registered.
    ///
    /// It searches for a registered add-on with the same id.
    ///
    /// * `addon`: add-on to check.
    pub fn is_addon_registered(&self, addon: &Addon) -> bool {
        self.find_registered_addon(&addon.id).is_some()
    }

    // Activates the product with the given code.
    fn activate_product(
        &mut self,
        zypp: &zypp_agama::Zypp,
        name: &str,
        version: &str,
        code: Option<&str>,
    ) -> RegistrationResult<()> {
        let product = Self::product_specification(name, version, self.arch);
        let mut params = self.connect_params.clone();
        params.token = code.map(ToString::to_string);

        tracing::debug!("Registering product {product:?}");
        let service = suseconnect_agama::activate_product(
            product,
            params,
            self.connect_params
                .email
                .as_ref()
                .map(|e| e.as_str())
                .unwrap_or(""),
        )?;

        if let Some(file) = Self::credentials_from_url(&service.url) {
            let path = self
                .root_dir
                .join(format!("etc/zypp/credentials.d/{}", file));
            tracing::debug!(
                "Creating the credentials file for {} at {}",
                &service.name,
                &path
            );
            suseconnect_agama::create_credentials_file(
                &self.creds.login,
                &self.creds.password,
                path.as_str(),
            )?;
            self.config_files.push(path);
        }

        // Add the libzypp service
        zypp.add_service(&service.name, &service.url)
            .map_err(|e| RegistrationError::AddService(service.name.clone(), e))?;
        let name = service.name.clone();
        self.services.push(service);
        zypp.refresh_service(&name)
            .map_err(|e| RegistrationError::RefreshService(name, e))?;
        Ok(())
    }

    /// Returns the registration information.
    ///
    /// It includes not only the basic data (like the registration code or the e-mail),
    /// but the list of extensions.
    pub fn to_registration_info(&self) -> RegistrationInfo {
        let addons: Vec<AddonInfo> = match self.base_product() {
            Ok(product) => product
                .extensions
                .into_iter()
                .map(|e| {
                    let registration = match self.find_registered_addon(&e.identifier) {
                        Some(addon) => AddonRegistration::Registered {
                            code: addon.code.clone(),
                        },
                        None => AddonRegistration::NotRegistered,
                    };

                    AddonInfo {
                        id: e.identifier,
                        version: e.version,
                        label: e.friendly_name,
                        available: e.available,
                        free: e.free,
                        recommended: e.recommended,
                        description: e.description,
                        release: e.release_stage,
                        registration,
                    }
                })
                .collect(),
            Err(error) => {
                tracing::error!("Failed to get the product from the registration server: {error}");
                vec![]
            }
        };

        RegistrationInfo {
            code: self.connect_params.token.clone(),
            email: self.connect_params.email.clone(),
            url: self.connect_params.url.clone(),
            addons,
        }
    }

    /// Writes to target system all registration configuration that is needed
    ///
    /// Beware that, if a certificate was imported, it is copied by the agama-security service.
    pub fn finish(&mut self, install_dir: &Utf8PathBuf) -> Result<(), RegistrationError> {
        suseconnect_agama::write_config(self.connect_params.clone())?;
        self.config_files
            .push(suseconnect_agama::DEFAULT_CONFIG_FILE.into());
        self.copy_files(install_dir)?;

        // FIXME: Copy services files. Temporarily solution because, most probably,
        // it should be handled by libzypp itself.
        if let Err(error) = copy_dir_all(
            self.root_dir.join("etc/zypp/services.d"),
            install_dir.join("etc/zypp/services.d"),
        ) {
            tracing::error!("Failed to copy the libzypp services files: {error}");
        };
        Ok(())
    }

    fn copy_files(&self, target_dir: &Utf8PathBuf) -> Result<(), RegistrationError> {
        for path in &self.config_files {
            let target_path = match path.strip_prefix(&self.root_dir) {
                Ok(relative_path) => target_dir.join(&relative_path),
                Err(_) => {
                    let relative_path = path.strip_prefix("/").unwrap_or(path);
                    target_dir.join(relative_path)
                }
            };
            tracing::info!("Copying credentials file {path} to {target_path}");
            std::fs::copy(path, target_path)
                .map_err(|e| RegistrationError::IO(path.to_string(), e))?;
        }

        Ok(())
    }

    fn base_product(&self) -> RegistrationResult<suseconnect_agama::Product> {
        let product = suseconnect_agama::show_product(
            self.base_product_specification(),
            self.connect_params.clone(),
        )?;
        Ok(product)
    }

    fn base_product_specification(&self) -> suseconnect_agama::ProductSpecification {
        Self::product_specification(&self.product, &self.version, self.arch)
    }

    fn product_specification(
        id: &str,
        version: &str,
        arch: Arch,
    ) -> suseconnect_agama::ProductSpecification {
        // We do not expect this to happen.
        suseconnect_agama::ProductSpecification {
            identifier: id.to_string(),
            arch: arch.to_string(),
            version: version.to_string(),
        }
    }

    fn credentials_from_url(url: &str) -> Option<String> {
        let url = Url::parse(url)
            .inspect_err(|e| tracing::warn!("Could not parse the service URL: {e}"))
            .ok()?;
        url.query_pairs()
            .find(|(k, _v)| k == "credentials")
            .map(|(_k, v)| v.to_string())
    }

    fn find_registered_addon(&self, id: &str) -> Option<&Addon> {
        self.addons.iter().find(|a| a.id == id)
    }
}

/// A builder for a [Registration] object.
///
/// It is used to configure the build a registration object. It allows to configure
/// the registration parameters like the product and version, the registration code,
/// the e-mail, etc.
/// [Registration] object.
#[derive(Debug)]
pub struct RegistrationBuilder {
    root_dir: Utf8PathBuf,
    product: String,
    version: String,
    code: Option<String>,
    email: Option<String>,
    url: Option<Url>,
}

impl RegistrationBuilder {
    /// Creates a new builder.
    ///
    /// It receives the mandatory arguments for registering a system.
    ///
    /// * `root_dir`: root directory where libzypp configuration lives.
    /// * `product`: product name (e.g., "SLES").
    /// * `version`: product version (e.g., "16.1").
    pub fn new(root_dir: Utf8PathBuf, product: &str, version: &str) -> Self {
        RegistrationBuilder {
            root_dir,
            product: product.to_string(),
            version: version.to_string(),
            code: None,
            email: None,
            url: None,
        }
    }

    /// Sets the registration code to use.
    ///
    /// * `code`: registration code.
    pub fn with_code(mut self, code: &str) -> Self {
        self.code = Some(code.to_string());
        self
    }

    /// Sets the e-mail associated to the registration.
    ///
    /// * `email`: registration e-mail.
    pub fn with_email(mut self, email: &str) -> Self {
        self.email = Some(email.to_string());
        self
    }

    /// Sets the URL of the registration server.
    ///
    /// * `url`: server URL.
    pub fn with_url(mut self, url: &Url) -> Self {
        self.url = Some(url.clone());
        self
    }

    /// Registers the system and return a [Registration] object.
    ///
    /// It announces the system, gets the credentials and registers the base product.
    ///
    /// * `zypp`: zypp instance.
    pub fn register(
        &self,
        zypp: &zypp_agama::Zypp,
        security_srv: &Handler<security::Service>,
    ) -> RegistrationResult<Registration> {
        let params = suseconnect_agama::ConnectParams {
            token: self.code.clone(),
            email: self.email.clone(),
            language: "en-us".to_string().into(),
            url: self.url.clone(),
            ..Default::default()
        };
        // https://github.com/agama-project/agama/blob/master/service/lib/agama/registration.rb#L294
        let version = self.version.split(".").next().unwrap_or("1");
        let arch = Arch::current().expect("Failed to determine the architecture");
        let target_distro = format!("{}-{}-{}", &self.product, version, arch.to_string());
        tracing::debug!("Announcing system {target_distro}");
        let creds = handle_registration_error(
            || suseconnect_agama::announce_system(params.clone(), &target_distro),
            security_srv,
        )?;

        // suseconnect_agama::announce_system(params.clone(), &target_distro)?;

        tracing::debug!(
            "Creating the base credentials file at {}",
            suseconnect_agama::GLOBAL_CREDENTIALS_FILE
        );
        suseconnect_agama::create_credentials_file(
            &creds.login,
            &creds.password,
            suseconnect_agama::GLOBAL_CREDENTIALS_FILE,
        )?;

        let mut registration = Registration {
            root_dir: self.root_dir.clone(),
            connect_params: params,
            product: self.product.clone(),
            version: self.version.clone(),
            arch,
            creds,
            services: vec![],
            addons: vec![],
            config_files: vec![suseconnect_agama::GLOBAL_CREDENTIALS_FILE.into()],
        };

        registration.activate_product(zypp, &self.product, &self.version, None)?;
        Ok(registration)
    }
}

/// Ancillary function to handle registration errors.
///
/// Runs the given function and handles potential SSL errors. If there is an SSL
/// error and it can be solved by importing the certificate, it asks the security
/// service whether to trust the certificate.
///
/// It returns the result if the given function runs successfully or return any
/// other kind of error.
fn handle_registration_error<T, F>(
    func: F,
    security_srv: &Handler<security::Service>,
) -> Result<T, RegistrationError>
where
    F: Fn() -> Result<T, suseconnect_agama::Error>,
{
    loop {
        let result = func();

        if let Err(suseconnect_agama::Error::SSL {
            code,
            message: _,
            current_certificate,
        }) = &result
        {
            if code.is_fixable_by_import() {
                let x509 = X509::from_pem(&current_certificate.as_bytes()).unwrap();
                match should_trust_certificate(&x509, security_srv) {
                    Ok(true) => {
                        if let Err(error) = suseconnect_agama::reload_certificates() {
                            tracing::error!("Could not reload the certificates: {error}");
                        }
                        continue;
                    }
                    Ok(false) => tracing::warn!("Do not trust the certificate"),
                    Err(error) => tracing::error!("Error processing the certificate: {error}"),
                }
            }
        }

        return Ok(result?);
    }
}

pub fn should_trust_certificate(
    certificate: &X509,
    security_srv: &Handler<security::Service>,
) -> Result<bool, security::service::Error> {
    // unwrap OK: unwrap is fine because, if we eat all I/O resources, there is not solution
    let rt = tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap();

    let security_srv = security_srv.clone();
    let certificate = certificate.clone();
    let handle = rt.spawn(async move {
        security_srv
            .call(security::message::CheckCertificate::new(
                certificate,
                "registration_server",
            ))
            .await
    });
    rt.block_on(handle).unwrap()
}
